nixos/modules/services/games/minecraft-server.nix

```diff
@@ -210,6 +239,21 @@ in {
           rm .declarative
         fi
       '');
+
+    preStop = optionalString (cfg.fifo != null) ''
+      ${concatMapStrings (message:
+        ''
+          echo "say ${message}" > ${cfg.fifo}
+          sleep ${cfg.warn.stopDelay}
+        ''
+      ) cfg.warn.stopMessage}
+    '';
+    };
+
+    systemd.sockets.minecraft-server = mkIf (cfg.fifo != null) {
+      description = "Minecraft stdin fifo file";
+      socketConfig.ListenFIFO = cfg.fifo;
+      wantedBy = [ "sockets.target" ];
```

Comment:
This FIFO should be changed to have minecraft's user and group with mode 0660 (only writable by user and group). This means people will have to add themselves to the minecraft group to be allowed to send commands to it. See the `SocketUser`, `SocketGroup` and `SocketMode` systemd options for doing this.